tests/Type/data/collection-group-by.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace CollectionGroupBy;
+
+use App\User;
+use Illuminate\Support\Collection;
+use function PHPStan\Testing\assertType;
+
+/**
+ * @param Collection $untyped
+ * @param Collection<int, User> $users
+ * @param Collection<string, User> $stringUsers
+ * @param Collection<string, string> $stringStrings
+ * @param 'id'|'name' $key
+ * @param bool $preserve
+ * @param array<int, string> $keys
+ */
+function test(
+    Collection $untyped,
+    Collection $users,
+    Collection $stringUsers,
+    Collection $stringStrings,
+    string $key,
+    bool $preserve
+): void {
+    assertType(
+        'Illuminate\Support\Collection<(int|string), Illuminate\Support\Collection<int, mixed>>',
+        $untyped->groupBy('id')
+    );
+    assertType(
+        'Illuminate\Support\Collection<(int|string), Illuminate\Support\Collection<int, mixed>>',
+        $untyped->groupBy(['id'])
+    );
+    assertType(
+        'Illuminate\Support\Collection<(int|string), Illuminate\Support\Collection<(int|string), mixed>>',
+        $untyped->groupBy('id', $preserve)
+    );
+    assertType(
+        'Illuminate\Support\Collection<int, Illuminate\Support\Collection<int, App\User>>',
+        $users->groupBy('id')
+    );
+    assertType(
+        'Illuminate\Support\Collection<int|string, Illuminate\Support\Collection<int, App\User>>',
+        $users->groupBy($key)
+    );
+    assertType(
+        'Illuminate\Support\Collection<int, Illuminate\Support\Collection<string, Illuminate\Support\Collection<int, App\User>>>',
+        $users->groupBy(['id', 'name'])
+    );
+    assertType(
+        'Illuminate\Support\Collection<int, Illuminate\Support\Collection<int<0, 1>, Illuminate\Support\Collection<int, App\User>>>',
+        $users->groupBy(['id', fn ($user) => rand(0, 1)])
+    );
+    assertType(
+        'Illuminate\Support\Collection<int, Illuminate\Support\Collection<string, App\User>>',
+        $stringUsers->groupBy('id', true)
+    );
+    assertType(
+        'Illuminate\Support\Collection<int, Illuminate\Support\Collection<int|string, App\User>>',
+        $stringUsers->groupBy('id', $preserve)
+    );
+    assertType(
+        'Illuminate\Support\Collection<int, Illuminate\Support\Collection<string, Illuminate\Support\Collection<int|string, App\User>>>',
+        $stringUsers->groupBy(['id', 'name'], $preserve)
+    );
+    assertType(
+        "Illuminate\Support\Collection<'', Illuminate\Support\Collection<int, string>>",
+        $stringStrings->groupBy(['id'])
+    );
+    assertType(
+        'Illuminate\Support\Collection<(int|string), Illuminate\Support\Collection<(int|string), App\User>>',
+        $users->groupBy($keys)
```

Comment:
if the `$keys` are unknown we defer to the phpdoc rather than trying to be smart, unfortunately leading to incorrect typing in many cases

the other option would be to return `Collection<(int|string), mixed>` but that also has downsides but is closer to the truth

Reply:
I would rather have less specific types than incorrect types---I think in this case we would be able to return `Collection<(int|string), Collection<(int|string), mixed>>` as `groupBy` always returns a collection of collections right?

Reply:
i'd have to check but yeah potentially, unless empty array causes it to return an empty top level collection? don't have anymore time for today will check tomorrow